### PR TITLE
subgraph rewriter supports matched pattern with no users

### LINKED ISSUE #143841: Subgraph rewriter failed when the matched pattern has no users
### 🐛 Describe the bug

The subgraph rewriter will throw an error "The returning_nodes should have at least one user node", when the matched pattern has no users in the original graph.
Can reproduce with below example

```python
class M(torch.nn.Module):
    def forward(self, x, y, cache):
        m = torch.mul(x, y)
        n = cache.index_copy(0, torch.tensor([0]), m)
        p = torch.ops.aten.copy.default(cache, n)
        q = torch.ops.aten.copy_.default(cache, p)
        u = torch.relu(cache)
        return u  # check the result to ensure cache is updated before relu op

def pattern(self_tensor, src_tensor):
    p = torch.ops.aten.copy.default(self_tensor, src_tensor)
    q = torch.ops.aten.copy_.default(self_tensor, p)
    return q

def replacement(self_tensor, src_tensor):
    q = torch.ops.aten.copy_.default(self_tensor, src_tensor)
    return q

def comparison(x, y, cache):
    m = torch.mul(x, y)
    n = cache.index_copy(0, torch.tensor([0]), m)
    q = torch.ops.aten.copy_.default(cache, n)
    u = torch.relu(cache)
    return u

traced = symbolic_trace(M())
print(traced)
comparison_fn = symbolic_trace(comparison)
print(comparison_fn)

subgraph_rewriter.replace_pattern(traced, pattern, replacement)
```

### Versions

Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov  6 2024, 20:22:13) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-126-generic-x86_64-with-glibc2.35
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        43 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               12
On-line CPU(s) list:                  0-11
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   1
Core(s) per socket:                   6
Socket(s):                            2
Stepping:                             0
BogoMIPS:                             4389.68
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xsaves arat pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    VMware
Virtualization type:                  full
L1d cache:                            384 KiB (12 instances)
L1i cache:                            384 KiB (12 instances)
L2 cache:                             12 MiB (12 instances)
L3 cache:                             71.5 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-11
Vulnerability Gather data sampling:   Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Mitigation; PTE Inversion; VMX flush not necessary, SMT disabled
Vulnerability Mds:                    Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.15.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.10.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu11==11.11.3.6
[pip3] nvidia-cuda-cupti-cu11==11.8.87
[pip3] nvidia-cuda-nvrtc-cu11==11.8.89
[pip3] nvidia-cuda-runtime-cu11==11.8.89
[pip3] nvidia-cudnn-cu11==8.7.0.84
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-curand-cu11==10.3.0.86
[pip3] nvidia-cusolver-cu11==11.4.1.48
[pip3] nvidia-cusparse-cu11==11.7.5.86
[pip3] nvidia-nccl-cu11==2.20.5
[pip3] nvidia-nvtx-cu11==11.8.86
[pip3] optree==0.12.1
[pip3] pytorch-triton==3.0.0+45fff310c8
[pip3] torch==2.6.0a0+gitf6cd540
[pip3] torchaudio==2.2.0.dev20240429+cu118
[pip3] torchvision==0.20.0.dev20240726+cpu
[conda] Could not collect

cc @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### PR DESCRIPTION
**Pull request by YangQun1**
Fixes #143841

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

### CODE REVIEW

**Comment by YangQun1** [comment]
Hi @ezyang , could you help to review? Many thanks.

File: torch/fx/subgraph_rewriter.py [hunk 1]
```diff
@@ -376,7 +378,21 @@ def _replace_pattern(
...
+            for n in reversed(original_graph.nodes):
+                if n in match.returning_nodes:
+                    first_next_node = next_node
+                    break
+                else:
+                    next_node = n
+        insert_point = (
+            first_user_node if first_user_node is not None else first_next_node
+        )
+        with original_graph.inserting_before(insert_point):  # type: ignore[possibly-undefined]
```
Comment (yushangdi) [inline torch/fx/subgraph_rewriter.py:395]: you can fix the linter error by adding something like `assert insert_point is not None`.

**Review by yushangdi — COMMENTED** [review]
The fix LGTM. Just a minor comment on silencing the linter error.

File: torch/fx/subgraph_rewriter.py (hunk 1 above)
Comment (YangQun1) [reply torch/fx/subgraph_rewriter.py]: Thanks for review. I added the assertion and removed the ignore

**Review by yushangdi — APPROVED** [review]

**Comment by YangQun1** [comment]
Hi @yushangdi , could you help to approve the CI workflow?

**Pull request closed without merging**